### ISSUE TITLE
More restrictive with `compositionLaw`, `inversion` or `equality`

### ISSUE THREAD

**Issue by xgbuils** [fibo/algebra-group#1]
Hi again, @fibo 

I'm working to throw TypeError when `compositionLaw`, `inversion` or `equality` receive or return values that are not contained in group. For example:

1.
If `compositionLaw`, `inversion` or `equality` receives arguments such that `contains` returns false, then it will be throw a TypeError with message:
```
algebra-group: "compositionLaw" must be called with arguments contained in group set
or
algebra-group: "inversion" must be called with arguments contained in group set
or
algebra-group: "equality" must be called with arguments contained in group set
```

2.
If `compositionLaw` or `inversion` return value such that `contains` returns false, then it will be throw a TypeError with message:
```
algebra-group: "compositionLaw" must return value contained in group set
or
algebra-group: "inversion" must return value contained in group set
```

3.
If `equality` return value such that `contains` returns false, then it will be throw a TypeError with message:
```
algebra-group: "inversion" must return boolean value
```

Does it seem like a good idea for you?

**Comment by fibo** [comment]
Hi @xgbuils, welcome  to *algebra-group*!

Yes, it is a really good idea.

In fact, the few error messages currently implemented were really useful when experimenting with groups other than real numbers.

I am sorry, there are no contribution guidelines written yet

but

if you clone the repo and run `npm install` it will set a git commit hook. If it does not complain when you commit, it is ok for me. Just tell me if you prefer to fork and do a PR, or just push on this repo directly.

**Comment by fibo** [comment]
Following your input, I updated also algebra-ring: https://github.com/fibo/algebra-ring/commit/5a2878157de8bf72f38e09e81a4977c28a722e90

Error messages are exported as constant strings (an idea stolen from an interesting [spf13](https://github.com/spf13) talk about Golang) so it is easier do manage errors.

**Comment by xgbuils** [comment]
Hi @fibo,

I can execute `npm test` after npm install. But I have problems executing `npm lint` or `npm check-deps`. Really if I execute `npm install` and does not complain, it is ok?

I'm ready to do the commit and PR but I want to ensure that everything is fine in local.

**Comment by xgbuils** [comment]
I've rewrite my first comment:

> 3.
> If `equality` return value such that `contains` returns false, then it will be throw a TypeError with message:
> ```
> algebra-group: "inversion" must return boolean value
> ```

I wanted to say:
3.
If `equality` return value **distinct to boolean**, then it will be throw a TypeError with message:
```
algebra-group: "inversion" must return boolean value
```

**Comment by fibo** [comment]
>  If it does not complain

I mean the commit hook (provided by [pre-commit](https://www.npmjs.com/package/pre-commit)). If you do `git commit`, the commit hook will run `npm run lint` which complains if code is not linted according to feross/standard.

If you want to commit anyway, there is no problem, it can be fixed later. You can bypass the commit hook by running `git commit -n`, then send the PR.

About the third point: `equality` should always return a boolean. It could be checked when group is built using `given.identity`, something like

```
var e = given.identity
if (give.equality(e, e) !== true) throw ...
```

In [algebra ring](https://github.com/fibo/algebra-ring) are passed two identities (**zero** and **one**) so there it could be possible to check also that `given.equality(zero, true)` is false.

I prefer to avoid a check on run time, I mean when equality is invoked.

**Comment by fibo** [comment]
published version `v0.5.0`
